docs/root/intro/arch_overview/http_connection_management.rst

```diff
@@ -111,6 +111,51 @@ previously attempted priorities.
 
 Envoy can be extended with custom retry plugins similar to how custom filters can be added.
 
+.. _arch_overview_internal_redirects:
+
+Internal redirects
+--------------------------
+
+Envoy supports "internal redirects", that is capturing 302-redirect responses internally,
+synthesizing a new request, sending it to the upstream specified by the new route match, and
+returning the redirected response as the response of the original request.
+
+Internal redirects are configured via the :ref:`internal redirect action
+<envoy_api_field_route.RouteAction.internal_redirect_action>`
+in the route configuration.
+
+Redirect handling is triggered by an upstream server sending a 302 response with an
+x-envoy-internal-redirect header containing the fully qualified URL (http://host/path) to redirect
+to. If the Envoy receiving this response is not configured for internal redirects, or is
+explicitly configured to REJECT, the 302 will be converted into a 500 and the
+x-envoy-internal-redirect will be removed, to prevent leaking potentially private URLs to untrusted
+clients. If the Envoy receiving the response is configured to PASS_THROUGH it will ignore the
+header and pass the response through, so that users with multi-level Envoy deployments can configure
+which layers of Envoy capture and handle the internal redirect. If the Envoy is configured to
+HANDLE, it will validate the redirect and modify the request headers as described below, and send
+the modified request through a new filter chain to a new upstream as selected by the route
+associated with the new URL.
+
+For a redirect to be successful it must pass the following checks
+1. Be a 302 response
+1. Have an x-envoy-internal-redirect with a valid, fully qualified URL matching the scheme of the original request.
+1. The request must have been fully processed by Envoy.
+1. The request must not have a body
+1. The request must have not been previously redirected, as determined by the presence of an x-envoy-original-url header
+Any failure will result in a 500 being sent downstream.
+
+Once the redirect has passed these checks, the request headers which were shipped to the original
+upstream will be modified by
+1. Putting the fully qualified original request URL in the x-envoy-original-url header
+2. Replacing the Authority/Host, Scheme, and Path headers with the values from x-envoy-internal-redirect header
+
+The altered request headers will then have a new route selected, be sent through a new filter chain,
+and then shipped upstream with all of the normal Envoy request sanitization taking place. Note that
+HTTP connection manager sanitization such as clearing untrusted headers will only be applied once,
+per-route header modifications will be applied on both the original route and the second route, even
+if they are the same, so be mindful configuring header modification rules to avoid duplicating
+undesired header values.
+
```

Comment:
Thanks for this nice writeup. Few thoughts on the design:
1. Why rely on the upstream being an envoy? why not make this generic such that Envoy will auto follow redirects from any upstream (HTTP 302 has the redirect URL). Typical deployments with Envoys at the edge are more likely to benefit from this option. Nginx has something similar I believe.
2. what is the idea behind the x-envoy-original-url header? Put another way, why rely on a header returned by an untrusted upstream envoy (x-envoy-original-url) instead of keeping track of the original url inside the caller Envoy itself? (may be this is what you meant but the docs are not clear).
3. The idea of rejecting a redirect and returning 500 seems a bit out of place in this "follow_redirects" option.  In my mind, `follow_redirects: on (aka handle) | off (aka passthrough,default)` is different from 'on receiving a response code XYZ, send 500 upstream'. The latter is what you are doing with on301,send500. Couldn't this be accomplished with a simpler disallow_redirects option which if true will cause envoy to send 500 automatically. 

I would suggest rephrasing the internal redirects thing to something like

```

Reply:
1) Any upstream can send this.  The header directs Envoy instances to do an internal redirect but there's no documented requirement the originator of the URL is an Envoy instance.  I'll try to make that more clear.

2) I think you misread the docs.
    Redirects are initiated by an upstream sending the redirect command.  Envoy then sends the x-envoy-original-url to the new upstream.  No upstream ever sends x-envoy-original-url, and we protect against untrusted downstreams sending it.    I'll add a workflow example 

3) The use case I am most familiar with is sending the first request to an auth server, then redirecting to a content URL to serve the response.  In this case it is often the case that is's unsafe to ever return the content URL to the user, as they can use it to bypass auth checks.   As such I'm inclined to leave both the defaults and the sanity checks as returning 500.  I could add a fail-open option (split handle into handle-or-fail or handle-or-pass-through) but realistically any failed internal redirect indicates a bug on the upstream or configuration and I'm not inclined to add extra complexity to more gracefully handle user failure :-P

Reply:
1. I was talking about 

>Redirect handling is triggered by an upstream server sending a 302 response with an
x-envoy-internal-redirect header containing the fully qualified URL (http://host/path) to redirect
to

piece in the docs above. Which seems to indicate that the upstream service needs to be aware of Envoy being the Downstream server instead of using `HTTP 302 <url>`

3. I dont disagree with the use case. What I am stating is that

> If the Envoy receiving this response is not configured for internal redirects, or is
explicitly configured to REJECT, the 302 will be converted into a 500 and the
x-envoy-internal-redirect will be removed, to prevent leaking potentially private URLs to untrusted
clients.

for the scenario above, isn't it sufficient to do the following:
1. `allow_redirects` which if set to true will cause envoy to accept and forward redirect responses to downstream. If set to false, will cause envoy to return a 500 to downstream without revealing any info about the redirect URL sent by the upstream.
2. `follow_redirects: on|off`. If allow_redirects is set to true and follow_redirects is on, then you follow any 302 response from upstream. If allow_redirects is true and follow_redirects is off, then you forward the 302 response as is to downstream. if allow_redirects is false, then follow_redirects setting will be ignored. Any 302 response from upstream will be turned into a HTTP 500 response (after scrubbing). It might be worthwhile to do (in follow ups) a `follow_redirect_depth: N` which controls the number of times we are willing to follow redirects (to avoid infinite loops, etc.).

Reply:
ahhh, that totally makes more sense, my bad.  So you'd prefer the originating proxy not need to be Envoy-aware, and Envoy intercept all 302s and use the location header?  That's doable and I'm interested to hear if folks think it's a more compelling use case.  My concern is twofold.  first, intercepting redirects in general is dangerous: often if you're doing cross-domain redirects you're not going to get the correct cookies, or may forward cookies or other PII from one server which should have them to another which is not. Between that and URL leakage concerns I'd rather have the upstream do 302 capturing which is opt-out (you have to do something special to get this behavior) and fail-closed (you don't leak urls by accident) rather than this being something that people turn up without a lot of thought in the hopes of getting a latency boost skipping a client RTT.
We can _almost_ achieve both both with an x-envoy-internal-redirect-fail-close header where default IR behavior was what you're asking for (intercepting 302), we pass-through by default or if #redirects > N (where N is 1 for now but we can always extend in a follow-up), but then the use case I care about doesn't work for multi-level proxying (if pass-through is default we can't allow an L2 envoy to pass a fail-closed URL to an L1 envoy without a configured option PASS_THROUGH_EVEN_FOR_FAIL_CLOSE which feels needlessly complicated)
Any thoughts on how to handle both cases gracefully?

Reply:
I'm inclined to agree with @rshriram that offering opt-in 302 following is going to be a compelling feature. @alyssawilk the reason you need to differentiate "IR" from "regular" is really because you share a route in which you want to do both? That makes sense to me. I think I would be inclined to just have 2 different config options for "regular" and for header driven "IR" with clear docs. I think this would be the easiest to understand. With that said, I don't feel very strongly about single config vs. split. @rshriram thoughts?

Reply:
between poor GH formatting and terminology mismatch, i am having a hard time decoding @alyssawilk 's use case for the multi-level envoy stuff.

> Between that and URL leakage concerns I'd rather have the upstream do 302 capturing which is opt-out (you have to do something special to get this behavior) and fail-closed (you don't leak urls by accident) rather than this being something that people turn up without a lot of thought in the hopes of getting a latency boost skipping a client RTT.
We can almost achieve both both with an x-envoy-internal-redirect-fail-close header where default IR behavior was what you're asking for (intercepting 302), we pass-through by default or if #redirects > N (where N is 1 for now but we can always extend in a follow-up), but then the use case I care about doesn't work for multi-level proxying (if pass-through is default we can't allow an L2 envoy to pass a fail-closed URL to an L1 envoy without a configured option PASS_THROUGH_EVEN_FOR_FAIL_CLOSE which feels needlessly complicated)
Any thoughts on how to handle both cases gracefully?

Would it be possible to scribble a diagram on paper napkin and stick it in slack so that its easier to visualize the actors involved?

Reply:
I think Matt's got it, and I agree with him that 2 configs is going to be less confusing than one.

I'll put together the change Matt and I thinks makes sense along with docs, and ask @rshriram to vet that they're substantially more clear than my github comments :-D